### PR TITLE
fix validator when we have false returned by the current element of the iterator

### LINKED ISSUE #40057: [DoctrineBridge] Error on UniqueEntityValidator with Mongodb and Doctrine
**Symfony version used**: 4.4.18
**doctrine/mongodb-odm version used**: 2.1.2
**doctrine/mongodb-odm-bundle version used** : 4.2.2

**Description**  
When we use the UniqueEntity validator of Symfony by the Unique constraint of the doctrine bundle with a repository method that return an Iterator (in my case an instance of `Doctrine\ODM\MongoDB\Iterator\CachingIterator` which implement `\Iterator` interface), Symfony does not use correctly the iterator.

An empty result arrive to [this line](https://github.com/symfony/symfony/blob/a05cc024d3d43cb84faea3969de8cc2e590e7a83/src/Symfony/Bridge/Doctrine/Validator/Constraints/UniqueEntityValidator.php#L150). The `current()` method return false instead of null so we end up with the result  `[false]` and the error message is created.

I tested the use case on a old version of Symfony (3.3) and doctrine (1.2.9 for the ODM and 3.5.3 for the bundle), Doctrine return a instance of `Doctrine\ODM\MongoDB\Cursor` class wich return null on the method `current()`

**How to reproduce**  
Add an `Doctrine\Bundle\MongoDBBundle\Validator\Constraints\Unique` constraint with a repository method that return an Iterator (for example with the `execute()` method of the QueryBuilder) and use the validator with an empty result.

**Possible Solution**  
Check also if the Iterator is valid like this : 
```
 $result = $result->valid() && null !== $result->current() ? [$result->current()] : [];
```

### PR DESCRIPTION
**Pull request by FabienSalles**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no 
| Deprecations? | no 
| Tickets       | Fix #40057
| License       | MIT

### CODE REVIEW

**Comment by FabienSalles** [comment]
@xabbuh my tests failed because of deprecation notices with my anonymous classes.

Why is it not possible ?
What changes should I made ?

Thanks in advance :).

**Comment by stof** [comment]
something looks weird in these deprecations. they look truncated.
@nicolas-grekas could it be caused by the NUL byte in the anonymous class name ? And if yes, is it in our code, in PHPUnit or in PHP ?

File: src/Symfony/Bridge/Doctrine/Tests/Validator/Constraints/UniqueEntityValidatorTest.php [hunk 1]
```diff
@@ -793,4 +793,85 @@ public function testValidateUniquenessCause()
...
+        $entity = new SingleIntIdEntity(1, 'foo');
+
+        return [
+            [$entity, new class() implements \Iterator {
+                public function current()
+                {
+                    return null;
+                }
+
+                public function valid()
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Doctrine/Tests/Validator/Constraints/UniqueEntityValidatorTest.php]: missing return type (same below)

**Review by nicolas-grekas — COMMENTED** [review]
here is how to fix the failure (which is quite cryptic right now, I'm going to submit a fix in a separate PR)

File: src/Symfony/Bridge/Doctrine/Tests/Validator/Constraints/UniqueEntityValidatorTest.php (hunk 1 above)
Comment (FabienSalles) [reply src/Symfony/Bridge/Doctrine/Tests/Validator/Constraints/UniqueEntityValidatorTest.php]: thanks, I fixed it !

**Review by stof — APPROVED** [review]

**Review by xabbuh — APPROVED** [review]

**Review by Nyholm — APPROVED** [review]
Great. Thank you

**Pull request merged**